File: 0442. Find All Duplicates in an Array.py
from typing import *

class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        res = []

        for x in nums:
            idx = abs(x) - 1 # 1-indexed

            if nums[idx] < 0:
                res.append(abs(x))
            else:
                nums[idx] = -nums[idx]
        
        return res

File: 001-500/0442. Find All Duplicates in an Array.java
// Medium
// Array, Hash Table
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/find-all-duplicates-in-an-array/

import java.util.*;

// Try to modify in place

class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        List<Integer> res = new ArrayList<>();

        for (int x : nums) {
            int index = Math.abs(x) - 1;

            // we can mark as negative means we visited already.
            if (nums[index] < 0) {
                res.add(Math.abs(x));
            } else {
                nums[index] = -nums[index];
            }
        }
        return res;
    }
}
